### PR TITLE
Add composite matched actions into filter_state

### PR DESCRIPTION
**Pull request by juanmolle**
Commit Message: Add composite matched actions into filter_state
Additional Description:

Macher logic in Composite extension could be very complex. It is useful to have the action name that match for each filter for logging in access log and be used for other custom filter. This PR add the information into filter_state the action matched for each composite filter. The struct will contains keys values, where the key is the filterConfigName and the value the action name that matched if it does.

Risk Level: low
Testing: yes
Docs Changes: yes
Release Notes: yes
Platform Specific Features: n/a

### CODE REVIEW

**Comment by juanmolle** [comment]
I would appreciate feedback of this feature before continue with the testing code.
This feature will be especially useful combined with this change #29678 when used composite, because allows to know exactly the delegated filter executed under the composite pattern.

**Comment by yanavlasov** [comment]
Adding @tyxia for the first pass review

**Review by tyxia — COMMENTED** [review]
Thanks a lot for working on this!
Yes, The issue that embedded filter's filter name is not visible is a TODO that we should address. I left a related [comment](https://github.com/envoyproxy/envoy/pull/30340#issuecomment-1776172235) before.

From high level, this PR seems to make sense, especially in terms of the logic that is added: expose `composite_action.typed_config().name` to composite filter. But the PR seems to be incomplete e.g., key function is not called. Could you please finish this PR and also add a composite filter integration test, so that it gives a completed view of the picture. Thanks! I will take another closer look after that.

**Comment by juanmolle** [comment]
I will need some help with the doc.

**Comment by juanmolle** [comment]
gentle ping @tyxia

File: source/extensions/filters/http/composite/filter.cc
Comment (tyxia) [inline source/extensions/filters/http/composite/filter.cc:50]: nit: the definition of  `setFilterAction` and `serializeAsString` function can just be added to the header file, like what you already did for `serializeAsProto`, as they are 1 line function body. This is a bit more consistent I think

**Review by tyxia — COMMENTED** [review]
Thanks for completing code and adding more test!

Mostly some nits.

File: source/extensions/filters/http/composite/action.h
Comment (juanmolle) [reply source/extensions/filters/http/composite/action.h:18]: If I pass as const & and it is created from a temporal, I will construct twice, the temporal and to create the member one, without taking advance to just move the object without recreated it. I am wrong?

File: source/extensions/filters/http/composite/filter.h
Comment (juanmolle) [reply source/extensions/filters/http/composite/filter.h:37]: 1 . The ACTION_NAME is the filter name to be apply
```
matchers:
                          - predicate:
                              single_predicate:
                                input:
                                  name: request-headers
                                  typed_config:
                                    "@type": type.googleapis.com/envoy.type.matcher.v3.HttpRequestHeaderMatchInput
                                    header_name: h1
                                value_match:
                                  safe_regex:
                                    google_re2: {}
                                    regex: value2
                            on_match:
                              action:
                                name: execute-filter-action-level-00002
                                typed_config:
                                  "@type": type.googleapis.com/envoy.extensions.filters.http.composite.v3.ExecuteFilterAction
                                  typed_config:
                                    name: **ACTION_NAME**
                                    typed_config:
                                      "@type": type.googleapis.com/envoy.extensions.filters.http.ext_proc.v3.ExternalProcessor
                                      grpc_service:
                                        envoy_grpc:
                                          cluster_name: httpbin-ext-authz-native-external-proc
                                        timeout: 5s
                                      processing_mode:
                                        request_header_mode: SEND
                                        response_header_mode: SEND
                                        request_trailer_mode: SKIP
                                        response_trailer_mode: SKIP
                                      message_timeout: 2s
```

2 - Yes, it can be consumed with access log and from a custom WASM on_log method, in this particular case the stringlified value is used, that is the reason to convert to json

```
access_log:
          - name: envoy.access_loggers.file
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog
              path: "/dev/stdout"
              typed_json_format: 
                status: "%RESPONSE_CODE%"
                jmo2: "%FILTER_STATE(envoy.extensions.filters.http.composite.matched_actions)%"
                jmo3: "%FILTER_STATE(envoy.filters.network.http_connection_manager.local_reply_owner:PLAIN)%"
```

File: source/extensions/filters/http/composite/action.h
Comment (tyxia) [reply source/extensions/filters/http/composite/action.h:18]: https://abseil.io/tips/117 this blog actually discussed those exact two options. It is right that no extra copy when copy elision kicks in with temporary string. But I think for the small string like here, the performance gain is neglectable .   

Quote from link above "That principle certainly applies to this technique: passing by const reference is simpler and safer, so it’s still a good default choice"

I don't have strong opinion here though. Either works for me.

File: source/extensions/filters/http/composite/filter.h
Comment (tyxia) [reply source/extensions/filters/http/composite/filter.h:37]: Thanks for details!

Yea, there was a typo in config example in my comment. Using this [name](https://github.com/envoyproxy/envoy/blob/7e834d7123f38e4f607b8fa94b13167f0992de1f/api/envoy/config/core/v3/extension.proto#L23) as ACTION_NAME sounds good to me. 

This is actually the name of embedded filter.

Comment (juanmolle) [reply source/extensions/filters/http/composite/filter.h]: Excelent, I have added 2 more test to check the 2 types of config, dynamic and typed after the merge from main

File: changelogs/current.yaml
Comment (tyxia) [inline changelogs/current.yaml:70]: Reword a bit?
```suggestion
    Added composite filter's matched action name to filter state with key
```

File: source/extensions/filters/http/composite/filter.h
Comment (tyxia) [inline source/extensions/filters/http/composite/filter.h:33]: tiny nit: I personally feel `MatchedActionInfo` is a better name as I am not sure what does`Type` refer to?

WDYT?

**Review by tyxia — APPROVED** [review]
LGTM modulo few tiny nits. Thank you for your contribution! 

Also defer to @yanavlasov for second pass and merge

File: changelogs/current.yaml
Comment (juanmolle) [reply changelogs/current.yaml:70]: Thanks, texts are difficult to me.

In addition I think we should add some reference here
https://github.com/envoyproxy/envoy/blob/main/docs/root/configuration/http/http_filters/composite_filter.rst
 but I don't know how to do it

Comment (tyxia) [reply changelogs/current.yaml]: You can update that file (docs/root/configuration/http/http_filters/composite_filter.rst) directly. You can find some examples in its file history https://github.com/envoyproxy/envoy/commits/main/docs/root/configuration/http/http_filters/composite_filter.rst

File: docs/root/configuration/http/http_filters/composite_filter.rst
Comment (tyxia) [inline docs/root/configuration/http/http_filters/composite_filter.rst:27]:
```suggestion
  * ``ACTION_NAME``: delegated filter name of the action that is matched.
```

**Review by tyxia — APPROVED** [review]
Thanks for adding the doc!

Just some nits/suggestions on documentation, otherwise LGTM

**Comment by juanmolle** [comment]
gentle ping @yanavlasov if you are not in PTO

File: docs/root/configuration/http/http_filters/composite_filter.rst
Comment (juanmolle) [reply docs/root/configuration/http/http_filters/composite_filter.rst:22]: thanks

File: docs/root/configuration/http/http_filters/composite_filter.rst
Comment (juanmolle) [reply docs/root/configuration/http/http_filters/composite_filter.rst:27]: thanks

**Comment by adisuissa** [comment]
@yanavlasov PTAL.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
